### PR TITLE
Using fault name instead of enum as the key of fault hash table

### PR DESCRIPTION
**Pull request by zhangh43**
GPDB fault injector uses fault enum as the key of fault hash table.
If someone wants to inject fault into gpdb extensions(a separate repo),
she has to hard code the extension related fault enums into gpdb core
code, this is not a good practice.
So we simply use fault name as the hash key to remove the need of hard
code the fault enum. Note that fault injector API doesn't change.

### CODE REVIEW

**Comment by zhangh43** [comment]
Concourse failed because gpaddon repo depends on the fault enum. I have created a pr on gpaddon repo to fix it. link: https://github.com/greenplum-db/gpaddon/pull/2 
PR in gpaddon will not be pushed before this PR is pushed, since it will break the gpdb_master pipeline.

@hlinnaka  @danielgustafsson  please review it.

File: src/include/utils/faultinjector.h [hunk 1]
```diff
@@ -332,7 +117,7 @@ typedef enum FaultInjectorState_e {
  */
 typedef struct FaultInjectorEntry_s {
 	
-	FaultInjectorIdentifier_e	faultInjectorIdentifier;
+	char	 faultInjectorIdentifier[NAMEDATALEN];
```
Comment (ashwinstar) [inline src/include/utils/faultinjector.h]: There are strings like `transaction_commit_pass2_from_aborting_create_needed_to_aborting_create` larger than NAMEDATALEN (64) ?

**Comment by zhangh43** [comment]
In the first PR, I remove the fault identifier enum, and change the fault input from identifier enum to fault name in 53 files.
On the second thought, I think we could keep the fault identifier enum in the fault entry data structure and just add a new field 'fault name'. Fault hash table use field 'fault name' instread of enum as the hash key.

**Comment by zhangh43** [comment]
@ashwinstar For non-existing faults (the ones not in identifier enum), we could use macro SIMPLE_FAULT_NAME_INJECTOR(FaultName) to inject faults in the code. The fault entries are stored in a shared hash table called faultInjectorShmem->hash. The hash key is the fault name. When the codes run to SIMPLE_FAULT_NAME_INJECTOR(MyFault), it will check whether the fault entry with name 'MyFault' is in the hash table.

**Comment by stanlyxiang** [comment]
+1

File: src/backend/utils/misc/faultinjector.c [hunk 2]
```diff
@@ -1244,7 +1256,9 @@ FaultInjector_NewHashEntry(
 	}
 		
 	entryLocal->faultInjectorType = entry->faultInjectorType;
-	
+	entryLocal->faultInjectorIdentifier = entry->faultInjectorIdentifier;
+	strncpy(entryLocal->faultName, entry->faultName, sizeof(entryLocal->faultName));
```
Comment (paul-guo-) [inline src/backend/utils/misc/faultinjector.c]: Need to set the last character to be NULL terminated (strncpy does not guarantee). Others look good.

Comment (zhangh43) [reply src/backend/utils/misc/faultinjector.c]: fixed

File: src/backend/postmaster/postmaster.c [hunk 3]
```diff
@@ -3289,9 +3289,10 @@ processTransitionRequest_faultInject(void * inputBuf, int *offsetPtr, int length
 	elog(DEBUG1, "FAULT INJECTED: Name %s Type %s, DDL %s, DB %s, Table %s, NumOccurrences %d  SleepTime %d",
 		 faultName, type, ddlStatement, databaseName, tableName, numOccurrences, sleepTimeSeconds );
 
+	strncpy(faultInjectorEntry.faultName, faultName, sizeof(faultInjectorEntry.faultName));
+	faultInjectorEntry.faultName[FAULT_NAME_MAX_LENGTH-1] = '\0';
```
Comment (danielgustafsson) [inline src/backend/postmaster/postmaster.c]: Better to use `strlcpy()` to get a guaranteed NULL terminated string.

File: src/backend/utils/misc/faultinjector.c (hunk 2 above)
Comment (danielgustafsson) [inline src/backend/utils/misc/faultinjector.c]: Same as above, better to use `strlcpy()`

**Comment by zhangh43** [comment]
fix to use strlcpy.

**Pull request merged**